// vscode-messenger-common/lib/messages.js
"use strict";
/******************************************************************************
 * Copyright 2022 TypeFox GmbH
 * This program and the accompanying materials are made available under the
 * terms of the MIT License, which is available in the project root.
 ******************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.isNotificationMessage = exports.isResponseMessage = exports.isRequestMessage = exports.isMessage = exports.BROADCAST = exports.HOST_EXTENSION = void 0;
exports.HOST_EXTENSION = { type: 'extension' };
exports.BROADCAST = { type: 'broadcast' };
function isMessage(obj) {
    return typeof obj === 'object' && obj !== null && typeof obj.receiver === 'object';
}
exports.isMessage = isMessage;
function isRequestMessage(msg) {
    return !!msg.id && !!msg.method;
}
exports.isRequestMessage = isRequestMessage;
function isResponseMessage(msg) {
    return !!msg.id && !msg.method;
}
exports.isResponseMessage = isResponseMessage;
function isNotificationMessage(msg) {
    return !msg.id && !!msg.method;
}
exports.isNotificationMessage = isNotificationMessage;
